openpype/hosts/substancepainter/api/pipeline.py

```diff
@@ -0,0 +1,347 @@
+        register_loader_plugin_path(LOAD_PATH)
+        register_creator_plugin_path(CREATE_PATH)
+
+        log.info("Installing callbacks ... ")
+        # register_event_callback("init", on_init)
+        self._register_callbacks()
+        # register_event_callback("before.save", before_save)
+        # register_event_callback("save", on_save)
+        register_event_callback("open", on_open)
+        # register_event_callback("new", on_new)
+
+        log.info("Installing menu ... ")
+        self._install_menu()
+
+        project_settings = get_current_project_settings()
+        self._install_shelves(project_settings)
+
+        self._has_been_setup = True
+
+    def uninstall(self):
+        self._uninstall_shelves()
+        self._uninstall_menu()
+        self._deregister_callbacks()
+
+    def has_unsaved_changes(self):
+
+        if not substance_painter.project.is_open():
+            return False
+
+        return substance_painter.project.needs_saving()
+
+    def get_workfile_extensions(self):
+        return [".spp", ".toc"]
+
+    def save_workfile(self, dst_path=None):
+
+        if not substance_painter.project.is_open():
+            return False
+
+        if not dst_path:
+            dst_path = self.get_current_workfile()
+
+        full_save_mode = substance_painter.project.ProjectSaveMode.Full
+        substance_painter.project.save_as(dst_path, full_save_mode)
+
+        return dst_path
+
+    def open_workfile(self, filepath):
+
+        if not os.path.exists(filepath):
+            raise RuntimeError("File does not exist: {}".format(filepath))
+
+        # We must first explicitly close current project before opening another
+        if substance_painter.project.is_open():
+            substance_painter.project.close()
+
+        substance_painter.project.open(filepath)
+        return filepath
+
+    def get_current_workfile(self):
+        if not substance_painter.project.is_open():
+            return None
+
+        filepath = substance_painter.project.file_path()
+        if filepath and filepath.endswith(".spt"):
+            # When currently in a Substance Painter template assume our
+            # scene isn't saved. This can be the case directly after doing
+            # "New project", the path will then be the template used. This
+            # avoids Workfiles tool trying to save as .spt extension if the
+            # file hasn't been saved before.
+            return
+
+        return filepath
+
+    def get_containers(self):
+
+        if not substance_painter.project.is_open():
+            return
+
+        metadata = substance_painter.project.Metadata(OPENPYPE_METADATA_KEY)
+        containers = metadata.get(OPENPYPE_METADATA_CONTAINERS_KEY)
+        if containers:
+            for key, container in containers.items():
+                container["objectName"] = key
+                yield container
+
+    @staticmethod
+    def create_context_node():
+        pass
```

Comment:
I am not sure this function is similar to the `set_context_settings`. If yes, I am not certain what to set up within the limited arguments as far as I read the doc of SP's python api
```
substance_painter.project.Settings(default_save_path: Optional[str] = None, normal_map_format: Optional[[NormalMapFormat](https://github.com/ynput/OpenPype/pull/4283/files#substance_painter.project.NormalMapFormat)] = None, tangent_space_mode: Optional[[TangentSpace](https://github.com/ynput/OpenPype/pull/4283/files#substance_painter.project.TangentSpace)] = None, project_workflow: Optional[[ProjectWorkflow](https://github.com/ynput/OpenPype/pull/4283/files#substance_painter.project.ProjectWorkflow)] = None, export_path: Optional[str] = None, default_texture_resolution: Optional[int] = None, import_cameras: Optional[bool] = None, mesh_unit_scale: Optional[float] = None)
```
If we really follow the API, only thing we may think is to set up the resolution, but I dont really know.
```suggestion
        substance_painter.project.Settings(default_texture_resolution=2048)
```